src/goto-harness/memory_snapshot_harness_generator.h

```diff
@@ -275,11 +275,68 @@ class memory_snapshot_harness_generatort : public goto_harness_generatort
   /// \return pointer depth of type \p t
   size_t pointer_depth(const typet &t) const;
 
-  /// Recursively test pointer reference
-  /// \param expr: expression to be tested
-  /// \param name: name to be located
-  /// \return true if \p expr refers to an object named \p name
-  bool refers_to(const exprt &expr, const irep_idt &name) const;
+  void collect_references(const exprt &expr, std::set<irep_idt> &reference_set)
+    const;
+
+  /// Simple structure for linearising posets. Should be constructed with a map
+  /// assigning a key to a set of keys that precede it.
+  template <typename Key, typename Value>
+  struct partial_ordert
+  {
+  public:
+    explicit partial_ordert(const std::map<Key, std::set<Key>> &por) : por(por)
+    {
+    }
+
+    void sort(
+      const std::vector<std::pair<Key, Value>> &input,
+      std::vector<std::pair<Key, Value>> &output)
+    {
+      std::map<Key, Value> searchable_input;
+      for(const auto &item : input)
+        searchable_input[item.first] = item.second;
+
+      for(const auto &item : input)
+        recurse_then_insert(item.first, searchable_input, output);
+    }
+
+  private:
+    const std::map<Key, std::set<Key>> &por;
+
+    std::set<Key> inserted;
+    std::set<Key> seen;
```

Comment:
Note that these make `sort` a destructive method, i.e. an instance becomes unusable after `sort` is first called. Might be better to just pass these on as parameters, or loopify the recursion and just keep these as local variables

Reply:
I wrapped the `dfs` in a `cleaner` function.